tools/PI/DevHome.PI/Controls/ExternalToolsManagementButton.cs

```diff
@@ -0,0 +1,280 @@
+        {
+            Text = _unpinMenuItemText,
+            Icon = GetFontIcon(CommonHelper.UnpinGlyph),
+            Tag = tool,
+        };
+        unPinMenuSubItemItem.Click += ExternalToolPinUnpin_Click;
+
+        var unRegisterMenuSubItemItem = new MenuFlyoutItem
+        {
+            Text = _unregisterMenuItemText,
+            Icon = GetFontIcon(_UnregisterButtonText),
+            Tag = tool,
+        };
+        unRegisterMenuSubItemItem.Click += UnregisterMenuItem_Click;
+
+        var menuSubItemFlyout = new MenuFlyout();
+
+        menuSubItemFlyout.Items.Add(pinMenuSubItemItem);
+        menuSubItemFlyout.Items.Add(unPinMenuSubItemItem);
+        menuSubItemFlyout.Items.Add(unRegisterMenuSubItemItem);
+
+        if (tool.IsPinned)
+        {
+            pinMenuSubItemItem.Visibility = Visibility.Collapsed;
+        }
+        else
+        {
+            unPinMenuSubItemItem.Visibility = Visibility.Collapsed;
+        }
+
+        menuItem.ContextFlyout = menuSubItemFlyout;
+
+        return menuItem;
+    }
+
+    private void ExternalToolItem_PropertyChanged(object? sender, PropertyChangedEventArgs e)
+    {
+        if (sender is not ExternalTool tool)
+        {
+            Debug.Assert(false, "What is this?");
+            return;
+        }
+
+        _dispatcher.TryEnqueue(() =>
+        {
+            MenuFlyout? flyout = this.ContextFlyout as MenuFlyout;
+            Debug.Assert(flyout is not null, "Why does this button not have a menu flyout?");
+
+            // Update the submenu item for this tool
+            foreach (MenuFlyoutItem menuItem in flyout.Items)
+            {
+                if (menuItem.Tag == tool)
+                {
+                    // Update the name if it's changed
+                    menuItem.Text = tool.Name;
+
+                    var imageIcon = new ImageIcon
+                    {
+                        Source = tool.ToolIcon,
+                    };
+
+                    menuItem.Icon = imageIcon;
+
+                    var menuSubItemFlyout = menuItem.ContextFlyout as MenuFlyout;
+                    Debug.Assert(menuSubItemFlyout != null, "Why is this null?");
+
+                    var pinSubItemItem = menuSubItemFlyout.Items[0] as MenuFlyoutItem;
+                    Debug.Assert(pinSubItemItem != null, "Why is this null?");
+
+                    var unPinSubItemItem = menuSubItemFlyout.Items[1] as MenuFlyoutItem;
+                    Debug.Assert(unPinSubItemItem != null, "Why is this null?");
+
+                    pinSubItemItem.Visibility = tool.IsPinned ? Visibility.Collapsed : Visibility.Visible;
+                    unPinSubItemItem.Visibility = tool.IsPinned ? Visibility.Visible : Visibility.Collapsed;
+                    break;
+                }
+            }
+        });
+    }
+
+    public void ExternalToolMenuItem_Click(object sender, RoutedEventArgs e)
+    {
+        if (sender is MenuFlyoutItem clickedMenuItem)
```

Comment:
NIT: `sender is MenuFlyoutItem clickedMenuItem && clickedMenuItemTag is ExternalTool tool`

Reply:
Thanks.